### ISSUE TITLE
[sklearn#33982] Improve `pairwise_distances` so that it works with array API inputs for all supported metrics

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#365]
**Mirrored from [scikit-learn#33982](https://github.com/scikit-learn/scikit-learn/issues/33982)**

Original title: Improve `pairwise_distances` so that it works with array API inputs for all supported metrics
Original labels: Array API
Created: 2026-05-11T14:37:43Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

The `pairwise_distances` function can't always be called when using array API. We should fix this.

`pairwise_distances` with `metric="mahalanobis"` (and other scipy-delegated metrics not in `PAIRWISE_DISTANCE_FUNCTIONS`) does not work with array API inputs on non-CPU devices.

The `"mahalanobis"` metric is not in scikit-learn's `PAIRWISE_DISTANCE_FUNCTIONS`, so `pairwise_distances` falls through to the scipy fallback path, which calls `scipy.spatial.distance.cdist` / `scipy.spatial.distance.pdist`. As of scipy 1.17.1, both functions call `np.asarray()` on their inputs, which fails for GPU tensors (e.g. PyTorch MPS/CUDA, CuPy).

I'm not sure if the solution for this is to implement this in scikit-learn or wait for scipy to add support for it. For other metrics we have implementations in Python in `sklearn/metrics/pairwise.py`, so maybe there is no performance downside to having a Python implementation? If yes, then this is "just" a question of turning the handle to implement new distances?
 
xref https://github.com/scikit-learn/scikit-learn/pull/33573/changes#r3079594482

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [44](https://github.com/godofecht/flow-scikit/issues/44) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [45](https://github.com/godofecht/flow-scikit/issues/45) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
